### PR TITLE
8278282: G1: Log basic statistics for evacuation failure

### PR DESCRIPTION
**Pull request by Hamlin-Li**
The original pr is at #6763 , which should be retired as we have decided to adjust part of optimization solution for evacuation failure (see #6627 for details), so the log will be adjusted accordiingly.
The basic log related to evacuation failed will looks like below based on this patch.

```
[13.126s][debug][gc,phases] GC(0)       Restore Retained Regions (ms): Min:  0.0, Avg: 197.4, Max: 1579.1, Diff: 1579.1, Sum: 1579.1, Workers: 8
[13.126s][debug][gc,phases] GC(0)         Evacuation Failure Regions:    Min: 1, Avg:  1.0, Max: 1, Diff: 0, Sum: 1, Workers: 1
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278282](https://bugs.openjdk.java.net/browse/JDK-8278282): G1: Log basic statistics for evacuation failure

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to e70dc665b90ff36d4117fb5337469ca95725ba18
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**)
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6860/head:pull/6860` \
`$ git checkout pull/6860`

Update a local copy of the PR: \
`$ git checkout pull/6860` \
`$ git pull https://git.openjdk.java.net/jdk pull/6860/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6860`

View PR using the GUI difftool: \
`$ git pr show -t 6860`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6860.diff">https://git.openjdk.java.net/jdk/pull/6860.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/gc/g1/TestGCLogMessages.java [hunk 1]
```diff
@@ -265,7 +265,7 @@ private void testConcurrentRefinementLogs() throws Exception {
     LogMessageWithLevel exhFailureMessages[] = new LogMessageWithLevel[] {
         new LogMessageWithLevel("Recalculate Used Memory", Level.DEBUG),
         new LogMessageWithLevel("Restore Preserved Marks", Level.DEBUG),
-        new LogMessageWithLevel("Remove Self Forwards", Level.DEBUG),
+        new LogMessageWithLevel("Restore Retained Regions", Level.DEBUG),
```
Comment (tschatzl) [inline test/hotspot/jtreg/gc/g1/TestGCLogMessages.java:268]: Please also check the work item in this test. Needs to be added just here.

**Review by tschatzl — APPROVED** [review]
Lgtm apart from the omission of that log message check.

I assume that it's just not worth adding log messages for the sub phases (sorting, ...) of the evacuation failure handling since we decided on the other option using the prev bitmap already. That's fine.
In any case we can always improve these messages.

File: test/hotspot/jtreg/gc/g1/TestGCLogMessages.java (hunk 1 above)
Comment (Hamlin-Li) [reply test/hotspot/jtreg/gc/g1/TestGCLogMessages.java:268]: Thanks Thomas, I have added checking newly added work items.

**Comment by Hamlin-Li** [comment]
Can I have another reviewer? Thanks

**Comment by Hamlin-Li** [comment]
Can I have another reviewer?
Thanks

**Comment by Hamlin-Li** [comment]
Can I have another reviewer?
Thanks

**Comment by Hamlin-Li** [comment]
Can I have another reviewer?
Thanks

File: src/hotspot/share/gc/g1/g1GCPhaseTimes.cpp [hunk 2]
```diff
@@ -132,6 +132,8 @@ G1GCPhaseTimes::G1GCPhaseTimes(STWGCTimer* gc_timer, uint max_gc_threads) :
   _gc_par_phases[MergePSS]->create_thread_work_items("LAB Waste", MergePSSLABWasteBytes);
   _gc_par_phases[MergePSS]->create_thread_work_items("LAB Undo Waste", MergePSSLABUndoWasteBytes);
 
+  _gc_par_phases[RestoreRetainedRegions]->create_thread_work_items("Evacuation Failure Regions:", RestoreRetainedRegionsNum);
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1GCPhaseTimes.cpp:135]: Based on the surrounding naming conventions, `RestoreRetainedRegionsNum` should probably be sth like `RestoreRetainedRegionsNumEvacFailRegions` or just the suffix `NumEvacFailRegions`.

**Review by albertnetymk — APPROVED** [review]
Just a minor comment.

**Review by walulyai — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1GCPhaseTimes.cpp (hunk 2 above)
Comment (Hamlin-Li) [reply src/hotspot/share/gc/g1/g1GCPhaseTimes.cpp:135]: Thanks Albert, I see your point.
But seems the surrounding naming convention is not that strict, and there is no uniform convention here. And to me, `RestoreRetainedRegionsNum` is more friendly to read and easily related to `RestoreRetainedRegions`, so I prefer to keep it as `RestoreRetainedRegionsNum`.

**Comment by Hamlin-Li** [comment]
Thanks @tschatzl @albertnetymk @walulyai for your reviews.

/integrate

**Pull request closed without merging**